hypertrace-graphql-label-application-rules-schema/src/main/java/org/hypertrace/graphql/label/application/rules/dao/LabelApplicationRuleResponseConverter.java

```diff
@@ -0,0 +1,334 @@
+    return ruleData.map(data -> new ConvertedLabelApplicationRule(rule.getId(), data));
+  }
+
+  private Optional<LabelApplicationRuleData> convertLabelApplicationRuleData(
+      org.hypertrace.label.application.rule.config.service.v1.LabelApplicationRuleData data) {
+    Optional<Action> action = convertAction(data.getLabelAction());
+    List<Condition> conditionList = convertCondition(data.getMatchingCondition());
+    if (conditionList.isEmpty() || action.isEmpty()) {
+      return Optional.empty();
+    }
+    return Optional.of(
+        new ConvertedLabelApplicationRuleData(data.getName(), conditionList, action.get()));
+  }
+
+  private Optional<Action.Operation> convertOperationInAction(
+      org.hypertrace.label.application.rule.config.service.v1.LabelApplicationRuleData.Action
+          action) {
+    switch (action.getOperation()) {
+      case OPERATION_MERGE:
+        return Optional.of(Action.Operation.OPERATION_MERGE);
+      default:
+        log.error("Unrecognized Operation type in Action{}", action.getOperation().name());
+        return Optional.empty();
+    }
+  }
+
+  private Optional<Action> convertAction(
+      org.hypertrace.label.application.rule.config.service.v1.LabelApplicationRuleData.Action
+          action) {
+    List<String> entityTypes = action.getEntityTypesList();
+    Optional<Action.Operation> operation = convertOperationInAction(action);
+    if (operation.isEmpty()) {
+      return Optional.empty();
+    }
+    switch (action.getValueCase()) {
+      case STATIC_LABELS:
+        StaticLabels staticLabels = convertStaticLabels(action.getStaticLabels());
+        return Optional.of(
+            new ConvertedAction(
+                entityTypes, operation.get(), staticLabels, null, Action.ValueType.STATIC_LABELS));
+      case DYNAMIC_LABEL_KEY:
+        return Optional.of(
+            new ConvertedAction(
+                entityTypes,
+                operation.get(),
+                null,
+                action.getDynamicLabelKey(),
+                Action.ValueType.STATIC_LABELS));
+      default:
+        log.error("Unrecognized Value type in Action {}", action.getValueCase().name());
+        return Optional.empty();
+    }
+  }
+
+  private StaticLabels convertStaticLabels(
+      org.hypertrace.label.application.rule.config.service.v1.LabelApplicationRuleData.Action
+              .StaticLabels
+          staticLabels) {
+    return new ConvertedStaticLabels(staticLabels.getIdsList());
+  }
+
+  private List<Condition> convertCondition(
+      org.hypertrace.label.application.rule.config.service.v1.LabelApplicationRuleData.Condition
+          condition) {
+    switch (condition.getConditionCase()) {
+      case LEAF_CONDITION:
+        Optional<Condition> convertedCondition =
+            convertLeafCondition(condition.getLeafCondition()).map(ConvertedCondition::new);
+        return convertedCondition.map(List::of).orElse(Collections.emptyList());
```

Comment:
Well, in that case what is the difference between an empty list and Optional.empty()? It would  be ambiguous right? In fact, in one of the previous PR's you have asked me to convert Optional list to normal list and indicate the broken case with an empty list.

Reply:
Yes, sorry for the confusion - the reality is it depends on the context, and it's a bit difficult to explain but let me give it a go - it's relative to the semantics and responsibility of the method and gets a bit at single responsibility.

I can't remember the specific situation where I asked about converting in the other way, but in a scenario where a method is asked to convert a list of conditions from one format to another, and may fail, then how do we distinguish a failed result from a correctly converted empty list result? An empty list may not be valid for the use case, but that's not the responsibility of a converter to decide - it just converts. It needs to indicate somehow that this is a successful conversion or not - and usually we either use an optional or throw an error to do that.

Now take another scenario, where, for example, we're building a request out of input arguments. There, our job is really to normalize the input. If a list of keys is an optional argument, the question is - does omitting that argument mean the same thing as setting it to an empty list? If yes, then we can normalize it and convert down to a list, always. If no, then we should again retain the optional.

An example of this distinction in action is actually well represented by proto. In proto, I can put a string field in a message, and if it's unset, then it just means the getter returns an empty string. If I change that proto message and mark that same string field as optional, its `getXyz()` semantics are identical, but now there's a `hasXyz()` method that allows the reader to distinguish between a default value and a completely unset value.